buildkite/run/action.yml

```diff
@@ -0,0 +1,97 @@
+---
+name: buildkite/run
+description: A GitHub Action for triggering a build on a Buildkite pipeline.
+inputs:
+  branch:
+    description: 'Branch the commit belongs to. This allows you to take advantage of your pipeline and step-level branch filtering rules.'
+    default: main
+    required: false
+  commit:
+    description: 'Ref, SHA or tag to be built.'
+    default: HEAD
+    required: false
+  env-vars:
+    description: 'Additional environment variables to set on the build.'
+    required: false
+  message:
+    description: 'The BK build message to be shown in the UI.'
+    default: "Triggered automatically - ${{ github.workflow_ref }} (${{ github.event_name }})"
+    required: false
+  org:
+    description: 'Buildkite org to interact with.'
+    default: elastic
+    required: false
+  pipeline:
+    description: 'Buildkite pipeline to interact with.'
+    required: true
+  token:
+    description: 'Buildkite token.'
+    required: true
+  wait-for:
+    description: 'Whether to wait for the build to finish.'
+    default: false
+    required: false
+
+outputs:
+  build:
+    description: "The Buildkite web build url"
+    value: ${{ steps.trigger-buildkite.outputs.build }}
+  number:
+    description: "The Buildkite build number"
+    value: ${{ steps.trigger-buildkite.outputs.number }}
+  state:
+    description: "The Buildkite build state"
+    value: ${{ steps.trigger-buildkite.outputs.state }}
+  url:
+    description: "The Buildkite build url"
+    value: ${{ steps.trigger-buildkite.outputs.url }}
+
+runs:
+  using: "composite"
+  steps:
+    - name: Trigger Buildkite pipeline
+      id: trigger-buildkite
+      run: |
+        ${{ github.action_path }}/run.sh \
+          '${{ env.BK_TOKEN }}' \
+          '${{ env.ORG }}' \
+          '${{ env.PIPELINE }}' \
+          '${{ env.ENV_VARS }}' \
+          '${{ env.MESSAGE }}' \
+          '${{ env.COMMIT }}' \
+          '${{ env.BRANCH }}'
```

Comment:
```suggestion
          "${BK_TOKEN}" \
          "${ORG}" \
          "${PIPELINE}" \
          "${ENV_VARS}" \
          "${MESSAGE}" \
          "${COMMIT}" \
          "${BRANCH}"
```

Reply:
I'm biased to use `env.` so I know the env variable is coming from GitHub itself, either a built-in variable or an explicit variable definition in the `env` section.

That's the rationale behind this.

What do you think?

Reply:
Is it just a convention for you?

Because I don't think it's true

If a previous step before this action exports an environment variable, it will also be accessible from `${{ env.XYZ }}`

e.g.

```
steps:
  - run: |
      echo "FOO=foo" >> $GITHUB_ENV
  - run: echo "${{ env.FOO }}"
```

also works, right?

Reply:
> Is it just a convention for you?
Yes, I made it up

>If a previous step before this action exports an environment variable, it will also be accessible from ${{ env.XYZ }}

If that's the case, it lands in my previous comment regarding 'I know the env variable is coming from GitHub'.


There could be something like:

```yaml
steps:
  - run: |
      foo
      bar
      delta ...
      ...
      MY_VAR='acme'
      ...
      ...
      ...
      ...
      echo "$MY_VAR"
      echo "${{ env.MY_OTHER_VAR }}"
```

If I see `$MY_VAR`, then I'd expect the variable is created within the shell step using a shell script format. While if I see `${{ env.MY_OTHER_VAR }}"` then I'd expect the variable is created on the fly using some GitHub actions mechanism (env, GITHUB_ENV or built-in).

Maybe I'm overthinking too much?

Reply:
Ah okay.. this is what you mean. Now I get it.

I know this is trusted input.. so it's fine. But this is still a potential injection point.
either the point itself or a future contribution using the same pattern unaware of the security issues.

Reply:
all good for now

Reply:
> I know this is trusted input.. so it's fine. But this is still a potential injection point.

Can you elaborate this?

IIUC, env is the approach to handle inputs in a secured manner, as stated in https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-an-intermediate-environment-variable

Reply:
Yes, but only if you use the language-specific environment variable

in bash it would be:

```yaml
- run: echo "$FOO"
  env: 
    FOO: "foo"
```

or in python shell

```yaml
- run: |
    println(os.environ['FOO'])
  env:
    FOO: "foo
  shell: python
```

but if you use `${{ env.XYZ }}` directly the yaml string is evaluated before the script is executed

```yaml
- run: |
    println('${{ env.FOO }}')
  env:
    FOO: "foo
  shell: python
```

This will become 

```yaml
- run: |
    println('foo')
  env:
    FOO: "foo
  shell: python
```

Reply:
> but if you use ${{ env.XYZ }} directly the yaml string is evaluated before the script is executed

I see! Then let's play safe as you pointed and skip my wrongly assumption for the sake of semantic :)